src/providers/wfs/qgswfsprovidermetadata.cpp

```diff
@@ -0,0 +1,384 @@
+    QDomElement valueReferenceElem = doc.createElement(
+                                       ( caps.version.startsWith( "2.0" ) ) ? filterPrefix + QStringLiteral( ":ValueReference" ) : filterPrefix + QStringLiteral( ":PropertyName" ) );
+    functionElem.appendChild( valueReferenceElem );
+    valueReferenceElem.appendChild( doc.createTextNode( geometryElement ) );
+    QDomElement literalElem = doc.createElement( filterPrefix + QStringLiteral( ":Literal" ) );
+    propertyIsEqualToElem.appendChild( literalElem );
+    literalElem.appendChild( doc.createTextNode( QStringLiteral( "false" ) ) );
+  }
+
+  return doc.toString();
+}
+
+
+QString QgsWfsProviderMetadata::suggestGroupNameForUri( const QString &uri ) const
+{
+  QgsWFSDataSourceURI wfsUri( uri );
+  return wfsUri.typeName();
+}
+
+
+QList<QgsProviderSublayerDetails> QgsWfsProviderMetadata::querySublayers( const QString &uri, Qgis::SublayerQueryFlags flags, QgsFeedback * /*feedback*/ ) const
+{
+  QList<QgsProviderSublayerDetails> res;
+
+  if ( ( flags & Qgis::SublayerQueryFlag::FastScan ) )
+    return res;
+
+  QgsWFSDataSourceURI wfsUri( uri );
+  if ( !wfsUri.isValid() )
+    return res;
+
+  QgsWfsCapabilities::Capabilities caps = QgsWFSProvider::getCachedCapabilities( uri );
+  if ( caps.version.isEmpty() )
+    return res;
+
+  QgsWFSProvider provider(
+    uri + " " + QgsWFSConstants::URI_PARAM_SKIP_INITIAL_GET_FEATURE + "='true'",
+    QgsDataProvider::ProviderOptions(), caps );
+  QgsProviderSublayerDetails details;
+  details.setType( QgsMapLayerType::VectorLayer );
+  details.setProviderKey( QgsWFSProvider::WFS_PROVIDER_KEY );
+  details.setUri( uri );
+  details.setName( wfsUri.typeName() );
+  details.setWkbType( provider.wkbType() );
+  res << details;
+
+  if ( wfsUri.hasGeometryTypeFilter() ||
+       !caps.supportsGeometryTypeFilters() )
+  {
+    if ( provider.wkbType() == QgsWkbTypes::Type::Unknown )
+      provider.issueInitialGetFeature();
+    details.setWkbType( provider.wkbType() );
+    return res;
+  }
+
+  if ( provider.wkbType() == QgsWkbTypes::Type::Unknown &&
+       provider.sharedData()->layerProperties().size() == 1 )
+  {
+    std::vector<std::unique_ptr<QgsWFSGetFeature>> requests;
+    std::set<QgsWFSGetFeature *> finishedRequests;
+    constexpr int INDEX_ALL = 0;
+    constexpr int INDEX_NULL = 1;
+    constexpr int INDEX_POINT = 2;
+    constexpr int INDEX_CURVE = 3;
+    constexpr int INDEX_SURFACE = 4;
+    // Order of strings in the list must be consistent with the INDEX_* enumeration above
+    const QStringList filterNames = { QString(), // all features
+                                      QString( "IsNull" ),
+                                      QStringLiteral( "IsPoint" ),
+                                      QStringLiteral( "IsCurve" ),
+                                      QStringLiteral( "IsSurface" )
+                                    };
+
+    const auto downloaderLambda = [&]()
+    {
+      QEventLoop loop;
```

Comment:
I'd capture feedback here and support early cancellation via it.

Reply:
done